pkgs/native_toolchain_c/lib/src/cbuilder/cbuilder.dart

```diff
@@ -195,3 +209,23 @@ class CBuilder extends CTool implements Builder {
     }
   }
 }
+
+extension on BuildConfig {
+  List<String> dynamicLinkingFlags(List<String> libraries) =>
+      switch (targetOS) {
+        OS.macOS || OS.iOS => [
+            '-L${outputDirectory.toFilePath()}',
+            for (final library in libraries) '-l$library',
+          ],
+        OS.linux || OS.android => [
+            '-Wl,-rpath=\$ORIGIN/.',
+            '-L${outputDirectory.toFilePath()}',
+            for (final library in libraries) '-l$library',
+          ],
+        OS.windows => [
+            for (final library in libraries)
+              outputDirectory.resolve('$library.lib').toFilePath(),
```

Comment:
Users might output their libs somewhere else (a subdir for example), so maybe it's better to have `List<Uri> libraries` and be explicit about file paths.

If we go for file paths, then we might need to look up the name of the library for macOS/iOS/Linux/Android, unless the compiler allows file paths as well (as opposed to directory and library name.) But for those ditto, if the user compiles dylibs into a subdirectory of the target dir, the linking fails.